src/backend/gl/src/tex.rs

```diff
@@ -220,10 +220,41 @@ pub fn format_to_glfull(format: NewFormat) -> Result<GLenum, ()> {
     })
 }
 
+fn determine_mipmap_level(desc: &t::Info, mip: Option<t::Mipmap>) -> t::Level {
+    use std::cmp::{min, max};
+    
+    match mip {
+        Some(t::Mipmap::Allocated) => {
+            fn mip_level1(w: u16) -> u8 {
+                ((w as f32).log2() + 1.0) as u8
+            }
+            fn mip_level2(w: u16, h: u16) -> u8 {
+                ((max(w, h) as f32).log2() + 1.0) as u8
+            }
+            fn mip_level3(w: u16, h: u16, d: u16) -> u8 {
+                ((max(w, max(h, d)) as f32).log2() + 1.0) as u8
+            }
+            match desc.kind {
+                t::Kind::D1(w) | t::Kind::D1Array(w, ..) => mip_level1(w),
+                t::Kind::D2(w, h, t::AaMode::Single) | t::Kind::D2Array(w, h, ..) => mip_level2(w, h),
+                t::Kind::D3(w, h, d) => min(desc.levels, mip_level3(w, h, d)),
+                t::Kind::Cube(w) | t::Kind::CubeArray(w, ..) => mip_level2(w, w),
+                _ => desc.levels,
+            }
+        },
+        _ => desc.levels
+    }
+}
+
 fn set_mipmap_range(gl: &gl::Gl, target: GLenum, (base, max): (u8, u8)) { unsafe {
     gl.TexParameteri(target, gl::TEXTURE_BASE_LEVEL, base as GLint);
     gl.TexParameteri(target, gl::TEXTURE_MAX_LEVEL, max as GLint);
 }}
+fn set_mipmap_creation_parameters(gl: &gl::Gl, target: GLenum) { unsafe {
```

Comment:
these are not creation parameters, it's the filtering. Also, not sure why you need this?